### PR TITLE
Fix/stlink update

### PR DESCRIPTION
**Pull request by rongcuid**
###### Motivation for this change
To update STLink utility.

Note that I really don't know how to keep my branch up to date with master, since I based it on my current system version. Nowhere in docs did I find any information about how I merge changes from master back. I tried `-s ours`, `-Xtheirs`, `-Xours`, `rebase --onto`, and none of them work.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [X] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by rongcuid** [comment]
I forgot to turn off verbose, and the udev rules aren't there. I need to fix those.

**Comment by rongcuid** [comment]
OK, I have to ask here... I don't have a clue about how I can make udev rules work. They are already in the `/etc/udev/rules.d` of the package, but they are not loaded, and I can't get them to load. I could really use some help here, but it is totally undocumented.

**Comment by FRidh** [comment]
@rongcuid 

if you haven't done so already, add https://github.com/NixOS/nixpkgs-channels as remote.
```
git remote add channels https://github.com/NixOS/nixpkgs-channels.git
git fetch channels
```
Checkout `nixpkgs-unstable`
```
git checkout channels/nixpkgs-unstable
```
Cherry-pick this commit you made.
```
git cherry-pick 604a2f2f1871fbdf2555f9ec4957f303c4d696e7
```
Save it as a new branch named `stlink-update2` (note the 2)
```
git checkout -b stlink-update2
```
And force push to this PR
```
git push -f origin stlink-update2:stlink-update
```

cc maintainer @bjornfor

**Comment by rongcuid** [comment]
I will test that. But I still want to know how to get those udev rules to work. They are not loaded by any means, despite being in `/nix/store/<pkg>/etc/udev/rules.d`

**Comment by FRidh** [comment]
I can't help you with that.

**Comment by bjornfor** [comment]
@rongcuid: If the udev rules are properly installed by the package (```$out/etc/udev/rules.d/```) you can enable them on your system with ```services.udev.packages = [ thepackage ]``` in ```/etc/nixos/configuration.nix```.

**Comment by rongcuid** [comment]
Ah, I see. I will push the new changes once I get back.

**Comment by rongcuid** [comment]
Um, tried to force push, experimented with stuff and accidentally deleted the branch... 

    git push -f origin stlink-update2:stlink-update

This doesn't seem to work. Is the syntax correct?

**Comment by rongcuid** [comment]
OK, it's the slash that causes problem. Trying to fix it

**Comment by rongcuid** [comment]
Ahh, after some `--no-ff` magic and now it should be fine. 

Note that I don't have time to test udev rules yet, but I am still interested in finding a way to inform the user to add the `services.udev` line in their config. Alternatively, I can write a module... But since I now shifted to `openocd`, I would probably not do that.

File: pkgs/development/tools/misc/stlink/default.nix [hunk 1]
```diff
@@ -1,30 +1,33 @@
...
     url = "https://github.com/texane/stlink/archive/${version}.tar.gz";
-    sha256 = "0b38a32ids9dpnz5h892l279fz8y1zzqk1qsnyhl1nm03p7xzi1s";
+    sha256 = "3e8cba21744d2c38a0557f6835a05189e1b98202931bb0183d22efc462c893dd";
   };
 
-  buildInputs = [ autoconf automake libtool pkgconfig libusb1 ];
-  preConfigure = "./autogen.sh";
+  buildInputs = [ cmake libusb1 ];
+  patchPhase = ''
+sed -i 's@/etc/udev/rules.d@$ENV{out}/etc/udev/rules.d@' CMakeLists.txt
```
Comment (FRidh) [inline pkgs/development/tools/misc/stlink/default.nix]: please indent correctly

File: pkgs/development/tools/misc/stlink/default.nix [hunk 2]
```diff
@@ -1,30 +1,33 @@
...
+sed -i 's@/etc/modprobe.d@$ENV{out}/etc/modprobe.d@' CMakeLists.txt
+'';
+  preInstall = ''
+mkdir -p $out/etc/udev/rules.d
```
Comment (FRidh) [inline pkgs/development/tools/misc/stlink/default.nix]: here as well

Comment (rongcuid) [inline pkgs/development/tools/misc/stlink/default.nix]: Um, is it possible to do the change directly here without touching my local repo?

Comment (FRidh) [inline pkgs/development/tools/misc/stlink/default.nix]: you could go to your fork and branch on GitHUb and use the edit button

**Comment by FRidh** [comment]
@bjornfor are these changes okay with out? Now the commits just need to be squashes, but we can do that.

**Comment by bjornfor** [comment]
@rongcuid: What is fixed in the first commit? If this is more than a version bump I'd like it documented in the commit message. (You can write that here and I can add it to the commit when squashing.)

**Comment by rongcuid** [comment]
Well, first that is what I was told by the Nixpkgs manual to do... Using `fix` prefix for a version bump.

I also took care of the switch from autotools to cmake.

**Comment by bjornfor** [comment]
@rongcuid: Hm, I never heard of that convention. Seems odd. Do you have a link? I removed the "fix/" prefix, squashed and pushed to master (9775a26da39e72298bae1914ddec24799f80e33f). Thanks!

Btw, I noticed a couple of things (for possible future work):
1. the pkgconfig file is installed with wrong includedir path. Not sure if it's upstream or us that are messing it up, but it ends up with a version suffix, but in the file system the include dir is version-less. Since this is a new feature (previous versions didn't ship pkgconfig file), I didn't think it was important enough to fix right now.
2. If cmake finds GTK, it'll build some sort of stlink GUI tool. Perhaps we want to add enable building that (optionally).

**Comment by rongcuid** [comment]
https://nixos.org/nixpkgs/manual/#idm140737318408432

**Comment by FRidh** [comment]
@rongcuid there's an example where the *branch* starts with `fix` (`$ git checkout -b 'fix/pkg-name-update'`). The branch name doesn't matter, that's just up to you as contributor. One can use this kind of notation to describe better what the branch contains.

**Comment by rongcuid** [comment]
Ah, ok.

**Pull request closed without merging**